notebook/nbextensions.py

```diff
@@ -299,118 +355,300 @@ def _config_file_name_default(self):
+        list=(ListNBExtensionsApp, "List nbextensions")
     )
 
     def start(self):
         super(NBExtensionApp, self).start()
 
         # The above should have called a subcommand and raised NoStart; if we
-        # get here, it didn't, so we should print a message.
+        # get here, it didn't, so we should self.log.info a message.
         subcmds = ", ".join(sorted(self.subcommands))
-        sys.exit("Please supply at least one subcommand: %s" % subcmds)
+        self.log.warn("Please supply at least one subcommand: %s" % subcmds)
+        sys.exit(1)
 
 main = NBExtensionApp.launch_instance
 
+#------------------------------------------------------------------------------
+# Private API
+#------------------------------------------------------------------------------
+
+
+def _should_copy(src, dest, verbose=1, log=None):
+    """Should a file be copied?"""
+    if not os.path.exists(dest):
+        return True
+    if os.stat(src).st_mtime - os.stat(dest).st_mtime > 1e-6:
+        # we add a fudge factor to work around a bug in python 2.x
+        # that was fixed in python 3.x: http://bugs.python.org/issue12904
+        if verbose >= 2:
+            log("%s is out of date" % dest)
+        return True
+    if verbose >= 2:
+        log("%s is up to date" % dest)
+    return False
+
+
+def _maybe_copy(src, dest, verbose=1, log=None):
+    """Copy a file if it needs updating."""
+    if log is None: log = print
+    if _should_copy(src, dest, verbose=verbose, log=log):
+        if verbose >= 1:
+            log("Copying: %s -> %s" % (src, dest))
+        shutil.copy2(src, dest)
+
+
+def _safe_is_tarfile(path):
+    """Safe version of is_tarfile, return False on IOError."""
+    try:
+        return tarfile.is_tarfile(path)
+    except IOError:
+        return False
+
+
+def _get_nbextension_dir(user=False, sys_prefix=False, prefix=None, nbextensions_dir=None):
+    """Return the nbextension directory specified"""
+    if sum(map(bool, [user, prefix, nbextensions_dir, sys_prefix])) > 1:
+        raise ArgumentConflict("cannot specify more than one of user, sys_prefix, prefix, or nbextensions_dir")
+    if user:
+        nbext = pjoin(jupyter_data_dir(), u'nbextensions')
+    elif sys_prefix:
+        nbext = pjoin(ENV_JUPYTER_PATH[0], u'nbextensions')
+    elif prefix:
+        nbext = pjoin(prefix, 'share', 'jupyter', 'nbextensions')
+    elif nbextensions_dir:
+        nbext = nbextensions_dir
+    else:
+        nbext = pjoin(SYSTEM_JUPYTER_PATH[0], 'nbextensions')
+    return nbext
+
+
+def _get_config_dir(user=False, sys_prefix=False):
+    if sum(map(bool, [user, sys_prefix])) > 1:
+        raise ArgumentConflict("Cannot specify more than one of user or sys_prefix")
+    if user:
+        nbext = jupyter_config_dir()
+    elif sys_prefix:
+        nbext = ENV_CONFIG_PATH[0]
+    else:
+        nbext = SYSTEM_CONFIG_PATH[0]
+    return nbext
+
+
+def _get_nbextension_metadata(package):
```

Comment:
Nick, great questions.

I thought a bit more about this...

The metadata we need here isn't in the package.json.

* The primary metadata we is the *location* in the Python package of the
JavaScript code. People could put npm packages anywhere in their python
package.
* Because of the typical way that people are using npm with Python
packages, they are often putting the pakage.json at the top-level of the
repo and then using webpack/browserify to copy the final build into the
Python package. Because of this, the package.json often isn't even in the
installed python package.

An example that shows both of these things is here:

https://github.com/jdfreder/jupyter-lightsaber

Summary: while I am 100% on board with us moving to npm, I don't think we
should go that way for the 4.2 nbextensions.

Cheers,

Brian



On Mon, Feb 1, 2016 at 11:06 AM, Nicholas Bollweg <notifications@github.com>
wrote:

> In notebook/nbextensions.py
> <https://github.com/jupyter/notebook/pull/879#discussion_r51464130>:
>
> > +    return nbext
> > +
> > +
> > +def _get_config_dir(user=False, sys_prefix=False):
> > +    if sum(map(bool, [user, sys_prefix])) > 1:
> > +        raise ArgumentConflict("Cannot specify more than one of user or sys_prefix")
> > +    if user:
> > +        nbext = jupyter_config_dir()
> > +    elif sys_prefix:
> > +        nbext = ENV_CONFIG_PATH[0]
> > +    else:
> > +        nbext = SYSTEM_CONFIG_PATH[0]
> > +    return nbext
> > +
> > +
> > +def _get_nbextension_metadata(package):
>
> I chimed in on this in the meeting, but perhaps starting to adopt the npm
> opinion now would be in order. Let's please not keep making new canonical
> locations and structures for the same string.
>
> Adopting this approach, once little extensions grow up and leave the nest,
> they already have all of their metadata ready to go, and could in turn be
> used by other kernels.
>
> Instead of _jupyter_nbextension_paths, perhaps
> _jupyter_nbextension_packages, where each of the paths returned must
> contain a package.json. In each of these, go ahead and claim the jupyter
> namespace, and have the dictionary that is currently returned by the python
> magic name instead be stored here.
>
> Here is an extreme way to encode all that information:
>
> # __init__.pydef _jupyter_nbextension_paths():
>     # yeah, cuz windows
>     return [
>         ["static", "some-extension"]
>     ]
>
> # package.json
> {
>   "name": "some-extension", # this is the install destination name
>   "version": "0.1.0",
>   "jupyter": {
>     "4": { # just be ready
>       "nbextension": {
>         "section": "notebook",
>         "src": ["."], # path bits, relative to this file
>         "require": "dist/index",
>       }
>     }
>   }
>   # ... all the other package stuff
> }
>
> Granted, this would be a bit weird for those of us that:
>
>    - have "mixed" repos that have package.json and setup.py at the same
>    level
>       - *stripping out package.json is a pretty gnarly anti-pattern as we
>       are starting to think about versioning*
>    -
>
>    use one package.json to configure multiple extensions
>    - *this is not the node way...*
>       - *but there's nothing stopping you from keeping this pattern, and
>       just having/building extra package.jsons*
>
>    Starting to reduce the amount of redundancy and freedom of choice (to
>    fail) in configuration would probably be worth it to me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/879/files#r51464130>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com